src/Azure/Orleans.ScheduledJobs.AzureStorage/AzureStorageJobShardManager.cs

```diff
@@ -0,0 +1,365 @@
+using System;
+using System.Collections.Concurrent;
+using System.Collections.Generic;
+using System.Threading;
+using System.Threading.Tasks;
+using Azure;
+using Azure.Storage.Blobs;
+using Azure.Storage.Blobs.Models;
+using Azure.Storage.Blobs.Specialized;
+using Microsoft.CodeAnalysis;
+using Microsoft.Extensions.Logging;
+using Microsoft.Extensions.Options;
+using Orleans.Runtime;
+
+namespace Orleans.ScheduledJobs.AzureStorage;
+
+public sealed partial class AzureStorageJobShardManager : JobShardManager
+{
+    private readonly BlobServiceClient _blobServiceClient;
+    private readonly string _containerName;
+    private BlobContainerClient _client = null!;
+    private readonly IClusterMembershipService _clusterMembership;
+    private readonly ConcurrentDictionary<string, AzureStorageJobShard> _jobShardCache = new();
+    private readonly ILogger<AzureStorageJobShardManager> _logger;
+
+    public AzureStorageJobShardManager(
+        BlobServiceClient client,
+        string containerName,
+        IClusterMembershipService clusterMembership,
+        ILogger<AzureStorageJobShardManager> logger)
+    {
+        _blobServiceClient = client;
+        _containerName = containerName;
+        _clusterMembership = clusterMembership;
+        _logger = logger;
+    }
+
+    public AzureStorageJobShardManager(
+        IOptions<AzureStorageJobShardOptions> options,
+        IClusterMembershipService clusterMembership,
+        ILogger<AzureStorageJobShardManager> logger)
+        : this(options.Value.BlobServiceClient, options.Value.ContainerName, clusterMembership, logger)
+    {
+    }
+
+    public override async Task<List<JobShard>> AssignJobShardsAsync(SiloAddress siloAddress, DateTimeOffset maxDateTime, CancellationToken cancellationToken = default)
+    {
+        await InitializeIfNeeded(cancellationToken);
+        LogAssigningShards(_logger, siloAddress, maxDateTime, _containerName);
+
+        var blobs = _client.GetBlobsAsync(traits: BlobTraits.Metadata, cancellationToken: cancellationToken);
```

Comment:
Should we use knowledge about the blob naming scheme here to perform a server-side filter ?

Reply:
Inline this into the await foreach to make it clearer that it's IAsyncEnumerable